### PR TITLE
CDAP-1185 Assign unique timestamps for non-transactional readless increments

### PR DESCRIPTION
**Pull request by ghelmling**
This change makes readless increments safe to use in a non-transactional context by assigning a region-unique timestamp to all writes, similar to how the Tephra TransactionManager assigns timestamp based transaction IDs.

The change breaks down into a few parts:
* IncrementHandlerState factored out to reduce code duplication between IncrementHandler implementations
* IncrementHandler implementations updated to assign timestamps in non-transactional case
* AbstractIncrementHandler pulled out to provide a single set of test cases for IncrementHandlerTest implementations.  Also added test cases for mixed increment, put, delete and TTL handling.
* HBaseTestBase, HBaseTest94, HBaseTest96 HBaseTest98 refactored to reduce code duplication (supports shared test cases in AbstractIncrementHandlerTest)

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/IncrementHandlerState.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    this.tableName = tableName;
+  }
+
+  @VisibleForTesting
+  public void setTimestampOracle(TimestampOracle timeOracle) {
+    this.timeOracle = timeOracle;
+  }
+
+  protected Supplier<TransactionStateCache> getTransactionStateCacheSupplier(String tableName, Configuration conf) {
+    String[] parts = tableName.split("\\.", 2);
```
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/IncrementHandlerState.java]: wanna add a comment about table name format?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/IncrementHandlerState.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  /**
+   * Returns the time-to-live (in milliseconds) for the given column family, transformed into the same precision
+   * used in assigning unique timestamps.
+   *
+   * @param familyName the column family name
+   * @return the time-to-live value
+   */
+  public long getFamilyTTL(byte[] familyName) {
+    Long configuredTTL = ttlByFamily.get(familyName);
+    return configuredTTL == null ? -1 : configuredTTL * TxConstants.MAX_TX_PER_MS;
```
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/IncrementHandlerState.java]: I may be wrong - but we may tell that ttl is in seconds. Or it is presented to user as sec, but internally in ms. Confused - I remember asked shankar to clean it, but not sure what it is right now. You may want to doublecheck

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/TimestampOracle.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+
+  /**
+   * Returns a timestamp value unique within the scope of this {@code TimestampOracle} instance.  For usage
+   * by HBase {@code RegionObserver} coprocessors, this normally means unique within a given region.
+   */
+  public long getUniqueTimestamp() {
+    long lastTs;
+    long nextTs;
+    do {
```
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/TimestampOracle.java]: hm - that would be a nice interview question ;)

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/increment/hbase/AbstractIncrementHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+      assertFalse(region.scanRegion(results, row));
+      // verify we have 100 individual cells, one per increment
+      assertEquals(100, results.size());
+      for (int i = 0; i < results.size(); i++) {
+        ColumnCell cell = results.get(i);
+        assertEquals(1L, Bytes.toLong(cell.getValue(), 2));
+        Assert.assertEquals((now - i - 1) * IncrementHandlerState.MAX_TS_PER_MS, cell.getTimestamp());
+      }
+
+      // verify that when summing we return the correct sum
```
Comment (abaranau) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/increment/hbase/AbstractIncrementHandlerTest.java]: the verification is missing or it goes below?

Comment (ghelmling) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/increment/hbase/AbstractIncrementHandlerTest.java]: The verification is below.  I can update the comment to say "during flushing".

**Comment by abaranau** [comment]
very minor comments. 

awesome reducing of code duplication!

LGTM

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/IncrementHandlerState.java (hunk 2 above)
Comment (ghelmling) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/IncrementHandlerState.java]: The REST APIs take a TTL in seconds, but internally (and on the table configuration) we store it in milliseconds.  DatasetInstanceHandler does this:

      long ttl = TimeUnit.SECONDS.toMillis(Long.parseLong
        (creationProperties.getProperties().get(OrderedTable.PROPERTY_TTL)));
      creationProperties.getProperties().put(OrderedTable.PROPERTY_TTL, String.valueOf(ttl));

Comment (ghelmling) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/increment/hbase/IncrementHandlerState.java]: Hmm, HBaseMetricsTableDefinition is not setting the TTL as a column family value though, it's setting the HBase native TTL setting.  Will need to update this.

**Comment by abaranau** [comment]
Good to merge!

**Comment by ghelmling** [comment]
Will merge as soon as https://builds.cask.co/browse/CDAP-RBT92-6 completes

**Pull request merged**